marquez_airflow/dag.py

```diff
@@ -211,6 +197,65 @@ def handle_callback(self, *args, **kwargs):
 
         return super().handle_callback(*args, **kwargs)
 
+    def _register_dataset(self, step_dataset, run_id=None):
+        inputs = None
+        if step_dataset:
+            self.register_datasets(step_dataset, run_id)
+            inputs = self._to_dataset_ids(step_dataset)
+        return inputs
+
+    def _begin_run_flow(self, step, location, start_date, end_date,
+                        run_args={}):
+        # Create a job to ensure it exists
+        self.get_or_create_marquez_client().create_job(
+            namespace_name=self.marquez_namespace,
+            job_name=step.name,
+            job_type=JobType.BATCH,
+            location=step.location or location,
+            input_dataset=self._register_dataset(step.inputs),
+            output_dataset=self._register_dataset(step.outputs),
+            context=step.context or {},
+            description=self.description)
+        self.log.info(start_date)
+        # Run must be called after job creation
+        run_id = self.create_run(step.name, run_args,
+                                 start_date,
+                                 end_date)
+        self.log.info(f"Created run: {run_id}")
+        return run_id
+
+    def _end_run_flow(self, step, location, run_id, actual_start_date,
+                      ti_state):
+        if actual_start_date:
+            self.get_or_create_marquez_client() \
+                .mark_job_run_as_started(run_id, actual_start_date)
+        # Recreate job to associate run
+        self.get_or_create_marquez_client().create_job(
+            namespace_name=self.marquez_namespace,
+            job_name=step.name,
+            job_type=JobType.BATCH,
+            location=step.location or location,
+            input_dataset=self._register_dataset(step.inputs),
+            output_dataset=self._register_dataset_if_success(step.outputs,
+                                                             run_id, ti_state),
+            context=step.context or {},
+            description=self.description,
+            run_id=run_id)
+
+    def _register_dataset_if_success(self, step_dataset, run_id, ti_state):
+        if ti_state in {State.SUCCESS, State.SKIPPED}:
+            return self._register_dataset(step_dataset, run_id)
+        else:
+            # If the task has failed, remove output dataset
+            return []
```

Comment:
Also codecov says it's not covered but we do have a test that covers the failed task bellow. What am I missing?

Reply:
The dataset would need to be registered so it can have a valid entry. I'm confused on what the correct logic should be. Do we just not associate the run_id with the output dataset?

Reply:
After a convo, we'll have it keep outputting the dataset when a job fails to aid root cause analysis